src/EditorFeatures/Test2/Simplification/CastSimplificationTests.vb

```diff
@@ -5035,6 +5036,128 @@ class C
             Await TestAsync(input, expected)
         End Function
 
+        <Fact, Trait(Traits.Feature, Traits.Features.Simplification)>
+        Public Async Function TestCSharp_DoNotSimplifyNullableGeneric() As Task
+            Dim input =
+<Workspace>
+    <Project Language="C#" CommonReferences="true">
+        <Document><![CDATA[
+#nullable enable
+using System.Threading.Tasks;
+class Program
+{
+    Task<string?> M()
+    {
+        string s1 = "test";
+        return {|Simplify:Task.FromResult<string?>|}(s1);
+    }
+}
+]]>
+        </Document>
+    </Project>
+</Workspace>
+
+            Dim expected =
+<code><![CDATA[
+#nullable enable
+using System.Threading.Tasks;
+class Program
+{
+    Task<string?> M()
+    {
+        string s1 = "test";
+        return Task.FromResult<string?>(s1);
+    }
+}
+]]>
+</code>
+
+            Await TestAsync(input, expected, csharpParseOptions:=TestOptions.Regular8WithNullableAnalysis)
+        End Function
+
+        <Fact, Trait(Traits.Feature, Traits.Features.Simplification)>
+        Public Async Function TestCSharp_DoNotSimplifyNullableComplexGeneric() As Task
+            Dim input =
+<Workspace>
+    <Project Language="C#" CommonReferences="true">
+        <Document><![CDATA[
+#nullable enable
+class Program
+{
+    void M()
+    {
+        string? s1 = null;
+        string s2 = {|Simplify:ValueOrDefault<string>|}(s1!, "hello");
+    }
+
+    static T ValueOrDefault<T>(T target, T defaultValue) where T : class? =>
+        target ?? defaultValue;
+}
+]]>
+        </Document>
+    </Project>
+</Workspace>
+
+            Dim expected =
+<code><![CDATA[
+#nullable enable
+class Program
+{
+    void M()
+    {
+        string? s1 = null;
+        string s2 = ValueOrDefault<string>(s1!, "hello");
+    }
+
+    static T ValueOrDefault<T>(T target, T defaultValue) where T : class? =>
```

Comment:
@333fred 	@jasonmalinowski 	 and I were discussing if this is indeed the right behavior, and what the `where T : class?` actually means.

Reply:
`where T : class?` means that `T` can be substituted with any class type, and that type could potentially be nullable. ie, both `string?` and `string` are valid substitutions, but `int?` and `int` are not.

Reply:
Compare that to `where T : class`. `string` is a valid substitution, but `string?` is not.

Reply:
For this ValueOrDefault what would it look like to write it where target _must_ be nullable but would return T. Would that just be `static T ValueOrDefault<T>(T? target, T? defaultValue) where T : class`?

Reply:
What was tripping me up with the particular scenario @ryzngard was adding didn't seem to be using the right modifiers, and so although the asserted behavior might be "right" the test wasn't making sense high level.

Reply:
So in this case, `s1!` is still `string?`, `string` still is valid for the constraint, and we need to specify `string` at the callsite because otherwise it would be `string?`. Then the `!` is needed to say "it's ok, I know what I'm doing" regardless of the flow analysis showing that the `ValueOrDefault` could not return a `null`?

Reply:
(note: just made an edit to my earlier question to remove an ? that would indeed have made it make no sense)

Reply:
```C#
string? s1 = null;
string s2 = M2(s1, string.Empty);

static T M2<T>(T? t, T defaultValue) where T : class
            => t ?? defaultValue;
```

That fit what you're talking about?